Test with infra-slot message latency

nfrisby opened the pull request:
This Draft PR is my first attempt at injecting networking latency into the `test-consensus` tests while simultaneously ensuring that every mini protocol message arrives in the same slot it was sent. The purpose is to permute the interleaving of various events without adding new difficult-to-predict mechanisms for Common Prefix violations.

This PR is partial progress towards Issue #229. This intermediate milestone is motivated by the test plan in (pending) PR #1128.

nfrisby commented:
Currently, this implementation only requires the CS and BF mini protocols to settle down. The transaction submission mini protocols never seem to quiet down -- if I wait for those mini protocol instances to settle down, the test execution live locks.

nfrisby commented:
It seems like a good idea to open a specific Issue for this PR, but it also seems like that should wait until PR 1128 is resolved.

nfrisby commented on ouroboros-consensus/test-consensus/Test/Dynamic/Network.hs:
```diff
@@ -133,6 +141,16 @@ runNodeNetwork registry testBtime numCoreNodes nodeJoinPlan nodeTopology
+      }
+    onSlotChange btime $ \s -> do
+        blockUntilQuiescent livePipesVar quiescenceThreshold
+        atomically $ writeTVar latestDoneSlot s
```
My main reasons for now:

  * `Test.Dynamic.General` is not the only call-site for `newTestBlockchainTime`
  * This current approach is pretty awkward, still. I'm not even sure "the slot length" is relevant anymore?
  * So I thought we'd figure it out at this call-site before I attempt to alter a function with multiple call-sites.

nfrisby replied:
The only reason I'm thinking of for leaving it in `runTestNetwork` is because it's free to introduce further details (beyond `latestDoneSlot` interactions) that `runNodeNetwork` does not care about, so defining `btime` outside seems right.

nfrisby commented on ouroboros-consensus/test-consensus/Test/Dynamic/Network.hs:
```diff
@@ -110,6 +114,8 @@ runNodeNetwork :: forall m blk.
                     , HasCallStack
                     )
                  => ResourceRegistry m
+                 -> DiffTime
```
I'm thinking a `data` type for the functions arguments.

nfrisby replied:
Resolved by adding the `NodeNetworkArgs` data type.

nfrisby commented on ouroboros-consensus/test-consensus/Test/Dynamic/Network.hs:
```diff
@@ -725,3 +753,81 @@ data MiniProtocolFatalException = MiniProtocolFatalException
+      sig2 <- get
+      if sig1 == sig2 then pure () else go sig2
+
+-- | Create a pipe backed by a 'LazyTMVar', add it to the live pipes, and
```
I only used `LazyTMVar` because that's what was being used. Since these mvars are simulating a network, it seems like making it strict would be in a sense more accurate. Though "time" is somewhat weird here due to `io-sim`, so I don't think it ultimately matters -- maybe only for tracing `error` calls perhaps?

I'll change to `StrictTMVar` and see if someone else objects.

mrBliss replied:
Ha, no, I think you're already using a `StrictTMVar`, since you're using `Control.Monad.Class.MonadSTM.Strict`. A partial type signature on `buffer` confirms this. So my comment only applies to the docstring :slightly_smiling_face:

nfrisby replied:
It now uses `TQueue`s. Though the function's comment is now pretty big and does not mention the exact data type.

nfrisby commented on ouroboros-consensus/src/Ouroboros/Consensus/BlockchainTime.hs:
```diff
@@ -178,7 +179,19 @@ newTestBlockchainTime
+    => ResourceRegistry m
+    -> NumSlots           -- ^ Number of slots
+    -> DiffTime           -- ^ Slot duration
+    -> (SlotNo -> m ())   -- ^ Blocks until slot is finished
```
Your sentence conveys a lot of the right intuition. However, I have a pedantic objection: one slot finishes exactly when the next starts, there's no duration in between.

I hope to rework this function (see other PR comments), but I will at the very least improve the names and comments.

nfrisby replied:
How about now?

nfrisby commented:
OK: I fixed bugs and squashed it down to a single commit. I still haven't addressed your prior comments. FYI I intend to improve the organization, but I wanted to push up the first version that seems to be correct even if not yet good :)

nfrisby commented:
I rebased onto master, improved the organization of the commits, simplified the code some, and commented it some. The `Test.Dynamic.*` tests passed 10,000 QuickCheck iterations as of 4e1180c8.

I still plan to break some of the new code (the "live pipes" stuff) out into its own module.

nfrisby commented on ouroboros-consensus/test-consensus/Test/Dynamic/Network.hs:
```diff
@@ -761,3 +828,152 @@ data MiniProtocolFatalException = MiniProtocolFatalException
+  , lpInFlight :: !(StrictTVar m Word64)
+  }
+
+-- | A variable mapping each live pipe to its own variable, which contains
```
This comment is out-of-date

nfrisby commented:
I believe these commits are ready for review.

```
* 9dc45513 - (HEAD -> nfrisby/test-infraslot-delays, origin/nfrisby/test-infraslot-delays) TOSQUASH comments (40 seconds ago) <Nicolas Frisby>
* 4a59748f - TODROP: anticipate Issue 1147 (8 minutes ago) <Nicolas Frisby>
* 93d15eb5 - test-consensus: add Issue 1147 repro (8 minutes ago) <Nicolas Frisby>
* 1d3ba876 - test-consensus: improve the shrinking in Test.Dynamic.General (8 minutes ago) <Nicolas Frisby>
* 7238a43d - test-consensus: add infra-slot delays (8 minutes ago) <Nicolas Frisby>
* 5ad27a46 - ouroboros-consensus: generalize newTestBlockchainTime (3 days ago) <Nicolas Frisby>
* d55cb13a - test-consensus: introduce NodeNetworkArgs (3 days ago) <Nicolas Frisby>
*   d3e71fac - (origin/master, origin/bors/staging, origin/HEAD) Merge #1122 (4 days ago) <iohk-bors[bot]>
```

nfrisby commented:
Yep, these tests revealed Issue 1147 :+1: 

Related: I re-requested review just now only to clear your Approval. I haven't made any changes yet, but the commits currently include a workaround for Issue 1147 that I don't belong on master. I'd switch the PR back to Draft status if I could.

nfrisby commented on ouroboros-consensus/test-consensus/Test/Dynamic/Util/LivePipes.hs:
```diff
@@ -0,0 +1,226 @@
+    inBuf     <- atomically Q.newTQueue
+    outBuf    <- atomically Q.newTQueue
+    liVar1SMG <- mapM split liSMG
+    void $ fork $ forever $ do
```
Done. I also used the `ResourceRegistry` instead of explicitly managing `PipeIds` in `Network.hs`.

nfrisby commented on ouroboros-consensus/test-consensus/Test/Dynamic/General.hs:
```diff
@@ -143,24 +169,34 @@ runTestNetwork ::
+      { nnaLatencySeed         = latencySeed
+      , nnaLatestReadySlot     = latestReadySlot
+      , nnaMaxLatencies    = MaxLatencies
+          { maxSendLatency = 10   -- io-sim "seconds"
```
I added and used

```
-- | Map a number of seconds to a 'DiffTime' planned for use in an @io-sim@
-- simulation context
--
-- The @io-sim@ layer can skip ahead through time, so this function tends to
-- have arguments with extreme magnitude.
--
ioSimSecondsToDiffTime :: Integer -> DiffTime
ioSimSecondsToDiffTime = fromInteger
```

The @io-sim@ aspect is so un-enforcable that a `newtype` seemed wrong.

nfrisby commented on ouroboros-consensus/test-consensus/Test/Dynamic/Network.hs:
```diff
@@ -201,10 +227,12 @@ runNodeNetwork NodeNetworkArgs
       -> (CoreNodeId, CoreNodeId)
       -> m ()
-    undirectedEdge tr nodeVars (node1, node2) = do
+    undirectedEdge tr mbSMG livePipesVar nodeVars (node1, node2) = do
```
Done -- that was vestigial from when the data type was `Maybe`

nfrisby commented:
OK, I think I addressed your latest comments and then some.

This PR is blocked by a proper fix for Issue 1147 -- its current commits contain a workaround. Once a fix is on master, I'll rebase (and squash) and then we can try to merge.

nfrisby commented:
Status: I just rebased and tweaked the commits. I think 0d2fa18 is mergeable, but the last commit is a "surgical" minimal-diff simplification of what Duncan is planning to do, if I understand.

nfrisby commented:
@dcoutts On my local copy of this PR, I reverted 0d2fa187, which is the old, demonstrative commit for the Issue 1147 bug, and experimentally added the following commit which you suggested this morning on Slack.

```
$ git show aefa2a4a
commit aefa2a4add464f7e330b0611f709c8e41cc1099d
Author: Nicolas Frisby <nick.frisby@iohk.io>
Date:   Mon Jan 6 07:01:09 2020 -0800

    new fix

diff --git a/ouroboros-network/src/Ouroboros/Network/BlockFetch/Decision.hs b/ouroboros-network/src/Ouroboros/Network/BlockFetch/Decision.hs
index 692c35a1..ab3db0e9 100644
--- a/ouroboros-network/src/Ouroboros/Network/BlockFetch/Decision.hs
+++ b/ouroboros-network/src/Ouroboros/Network/BlockFetch/Decision.hs
@@ -920,7 +920,7 @@ fetchRequestDecision FetchDecisionPolicy {
              inFlightBytesLowWatermark
              inFlightBytesHighWatermark
 
-  | peerFetchReqsInFlight == 0
+  | peerFetchStatus == PeerFetchStatusReady Set.empty
   , let maxConcurrentFetchPeers = case fetchMode of
                                     FetchModeBulkSync -> maxConcurrencyBulkSync
                                     FetchModeDeadline -> maxConcurrencyDeadline
```

The 1147 repro failed :( I haven't yet investigated why -- just FYI.

dcoutts commented:
Thanks! I still think my basic approach should work, but I noticed a probable bug in my solution. I've updated branch `dcoutts/issue-1147`. I think it's worth checking if that solves these failure cases.

nfrisby commented:
> Thanks! I still think my basic approach should work, but I noticed a probable bug in my solution. I've updated branch `dcoutts/issue-1147`. I think it's worth checking if that solves these failure cases.

It fails with the same trace as in my previous comment :( Here's what I tested with:

```
$ git cherry-pick f3d43680 924e818e a269cfcf
[nfrisby/test-infraslot-delays 634f8bc6] Fix for block fetch decline decisions not being revisited appropriately
 Author: Duncan Coutts <duncan@well-typed.com>
 Date: Mon Jan 6 13:50:57 2020 +0000
 1 file changed, 1 insertion(+), 1 deletion(-)
[nfrisby/test-infraslot-delays 59796bc2] Improve comment doc on PeerFetchStatusReady constructor
 Author: Duncan Coutts <duncan@well-typed.com>
 Date: Mon Jan 6 13:54:40 2020 +0000
 1 file changed, 4 insertions(+)
[nfrisby/test-infraslot-delays 737716b1] Try and fix it properly this time.
 Author: Duncan Coutts <duncan@well-typed.com>
 Date: Fri Jan 10 10:05:48 2020 +0000
 2 files changed, 12 insertions(+), 2 deletions(-)
$ git lg -10
* 737716b1 - (HEAD -> nfrisby/test-infraslot-delays) Try and fix it properly this time. (31 seconds ago) <Duncan Coutts>
* 59796bc2 - Improve comment doc on PeerFetchStatusReady constructor (31 seconds ago) <Duncan Coutts>
* 634f8bc6 - Fix for block fetch decline decisions not being revisited appropriately (31 seconds ago) <Duncan Coutts>
* 3d459374 - DO NOT MERGE tracing for advising Duncan (55 seconds ago) <Nicolas Frisby>
* 9f2e1483 - DISCARD narrow tests (4 days ago) <Nicolas Frisby>
* 16f4ca7e - Revert "TODROP: anticipate Issue 1147" (4 days ago) <Nicolas Frisby>
* 0d2fa187 - (origin/nfrisby/test-infraslot-delays) TODROP: anticipate Issue 1147 (9 weeks ago) <Nicolas Frisby>
$
```

I just pushed 3d459374 to `origin/nfrisby/issue-229-issue-1147-for-duncan`. Does that build locally for you, @dcoutts? Because of 9f2e1483, `test-consensus` only runs the 1147 repro, and because of 3d459374 it dumps a bunch of trace events to `stderr`.

nfrisby commented:
I'm Closing this PR. The upcoming work for Praos is higher priority and will supersede it.

Status: PR #1705 resolved the BlockFetch bug that this PR had revealed (Issue #1147).
